sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java

```diff
@@ -113,12 +117,148 @@ static Builder builder() {
     Method targetMethod();
   }
 
-  /** A type of optional parameter of the {@link DoFn.ProcessElement} method. */
-  public enum Parameter {
-    BOUNDED_WINDOW,
-    INPUT_PROVIDER,
-    OUTPUT_RECEIVER,
-    RESTRICTION_TRACKER
+  /** A descriptor for an optional parameter of the {@link DoFn.ProcessElement} method. */
```

Comment:
This is a design discussion not yet concluded. We will need to provide some per-window callbacks. The one that is absolutely mandatory is `@OnWindowExpiration` to flush any state left when the window is being GC'd. I can see utility for per-key-and-window versions of `StartBundle` and `FinishBundle` or other lifecycle methods so that they can similarly initialize and flush state at the associated granularity. I do not have a firm conclusion about which of these methods delivers the most value relative to the restrictions it places on execution plans, or which might be completely necessary.